R/utils.R

```diff
@@ -140,3 +140,107 @@ edit.data.table = function(name, ...) {
   setDT(NextMethod('edit', name))[]
 }
 # nocov end
+
+col_helper = function(x, colsub, origin_call = NULL) {
+  ## helper that could replace some common code used for by, .SDcols, j, duplicated, unique, dcast, 
+  ## or anything that looks for columns
+  
+  #' @param x required; a data.frame, data.table, tibble, or list.
+  #' @param colsub takes the same arguments as .SDcols such as
+  #' `patterns()`, `var1:var2`, a variable, or a vector of characters.
+  #' When TRUE, will return seq_len(length(x))
+  #' @param origin_call takes character vector of length 1. In the short-term
+  #' there will be different modes based on the call. In the long-term, it's just
+  #' to help the end-user what caused the error
+  #' @return Integer vector with an attribute of whether negation was in the call
+  
+  # see @jangorecki #4174 for how to get the vast majority of this into C
+  
+  x_len = length(x)
+  
+  if (length(colsub) == 1L && isTRUE(colsub)){
+    ansvals = seq_len(x_len)
+    attr(ansvals, "negate") = FALSE
+    return(ansvals)
+  }
+  
+  # FR #4979 - negative numeric and character indices for SDcols
+  # fix for #5190. colsub[[1L]] gave error when it's a symbol.
+  negate_cols = is.call(colsub) && deparse(colsub[[1L]], 500L, backtick=FALSE) %in% c("!", "-")
```

Comment:
we could use `%iscall%` here?

Reply:
Yes! I kept the original blindly in there - who knows what's up with the ```500L``` and ```backtick=FALSE)```.  

Do we care about microsecond performance? ```%iscall%``` is generally slower than ```is.call(e) && e[[1L]] == "something"```. But on the bright side, it looks great.

```r
## looks great
  colsub %iscall% c("!", "-")
## is faster
  is.call(colsub) && (colsub[[1L]] == "!" || colsub[[1L]] == "-")

e = quote(1:3)
microbenchmark::microbenchmark(
  is.call(e) && (e[[1L]] == ":" || e[[1L]] == "!"),
  data.table:::'%iscall%'(e, c(":", "!")),
  is.call(e) && as.character(e[[1L]]) %in% c(":", "!")
)
Unit: nanoseconds
                                                 expr  min   lq  mean median  
     is.call(e) && (e[[1L]] == ":" || e[[1L]] == "!")  900 1100  1681   1400  
              data.table:::"%iscall%"(e, c(":", "!")) 6400 6900 10446   7650 
 is.call(e) && as.character(e[[1L]]) %in% c(":", "!") 2000 2200  3269   2400

Reply:
in this benchmark you also include `:::` call, which IMO should not be included

Reply:
after defining `%iscall%` in scope:

```
Unit: nanoseconds
                                                 expr  min   lq     mean median     uq     max neval cld
     is.call(e) && (e[[1L]] == ":" || e[[1L]] == "!")  724  788  1037.61  817.5  867.0   12307   100   a
                               e %iscall% c(":", "!") 1661 1738 23163.82 1793.0 1852.5 2123262   100   a
 is.call(e) && as.character(e[[1L]]) %in% c(":", "!") 1319 1424  1636.51 1500.5 1607.0   10766   100   a
```

Reply:
these are nonoseconds so it I thing those numbers may vary a lot on different machines

Reply:
increasing `neval`

```
Unit: nanoseconds
                                                 expr  min   lq      mean median   uq       max neval cld
     is.call(e) && (e[[1L]] == ":" || e[[1L]] == "!")  461  545  823.7597    580  626 169334249 1e+06  a 
                               e %iscall% c(":", "!") 1050 1206 1449.7929   1289 1380  23513043 1e+06   b
 is.call(e) && as.character(e[[1L]]) %in% c(":", "!")  819  989 1178.1340   1060 1158   2651089 1e+06   b
```

Here, it's not fair to use `e = quote(1:3)` since we _always_ stop out at the first part of `||`, here's the timing switching the order:

```
Unit: nanoseconds
                                                 expr  min   lq     mean median   uq       max neval cld
     is.call(e) && (e[[1L]] == "!" || e[[1L]] == ":")  760  854 1049.941    888  941  23475615 1e+06   a
                               e %iscall% c(":", "!") 1039 1184 1480.185   1230 1312  10060629 1e+06   a
 is.call(e) && as.character(e[[1L]]) %in% c(":", "!")  822  967 1498.495   1024 1100 277516160 1e+06   a
```

Lastly, `%in%` is not optimized for `character` input, which is why we made `%chin%`:

```
Unit: nanoseconds
                                                   expr  min   lq     mean median   uq      max neval cld
       is.call(e) && (e[[1L]] == "!" || e[[1L]] == ":")  759  853 1037.578    896  968 19875883 1e+06 a  
                                 e %iscall% c(":", "!") 1024 1178 1370.742   1233 1329 10193437 1e+06   c
 is.call(e) && as.character(e[[1L]]) %chin% c(":", "!")  805  945 1143.439    998 1077 24712648 1e+06  b 
```

Most of the gap appears to be due to the overhead of using `%chin%` instead of its code directly:

```
Unit: nanoseconds
                                             expr  min   lq     mean median   uq       max neval cld
 is.call(e) && (e[[1L]] == "!" || e[[1L]] == ":")  762  861 1186.890    899  966 195036437 1e+06   a
                           e %iscall% c(":", "!") 1030 1176 1383.657   1230 1316  10084976 1e+06   a
         is.call(e) && e[[1L]] %chin% c(":", "!")  724  842 1025.460    889  959  26607578 1e+06   a
```

Reply:
are you getting similar results when trying on a different machine?

Reply:
I stand corrected, thanks! So when you do ```cc()```, you can just do ```e %iscall% ":"``` without any additional hurdles? I only used ```:::``` because ```%iscall%``` is not exported.

Reply:
actually i just defined it by hand, but yes, `cc()` should accomplish the same IINM